hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/ReadHdfsNewApiP.java

```diff
@@ -0,0 +1,289 @@
+        this.trav = traverseIterable(splits)
+                .flatMap(this::traverseSplit);
+        this.projectionFn = projectionFn;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        HazelcastInstanceImpl instance = (HazelcastInstanceImpl) context.jetInstance().getHazelcastInstance();
+        serializationService = instance.getSerializationService();
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(trav);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Traverser<R> traverseSplit(InputSplit split) {
+        RecordReader<K, V> reader;
+        try {
+            TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(configuration, new TaskAttemptID());
+            reader = inputFormat.createRecordReader(split, attemptContext);
+            reader.initialize(split, attemptContext);
+        } catch (IOException | InterruptedException e) {
+            throw sneakyThrow(e);
+        }
+
+        return () -> {
+            try {
+                while (reader.nextKeyValue()) {
+                    // we clone the key/value if not configured to reuse the object because some
+                    // of the record-readers return the same object for `reader.getCurrentKey()`
+                    // and `reader.getCurrentValue()` which is mutated for each `reader.nextKeyValue()`
+                    K key = cloneIfNecessary(reader.getCurrentKey());
+                    V value = cloneIfNecessary(reader.getCurrentValue());
+                    R projectedRecord = projectionFn.apply(key, value);
+                    if (projectedRecord != null) {
+                        return projectedRecord;
+                    }
+                }
+                reader.close();
+                return null;
+            } catch (Exception e) {
+                throw sneakyThrow(e);
+            }
+        };
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> T cloneIfNecessary(T o) throws IOException, IllegalAccessException, InstantiationException {
+        if (configuration.getBoolean(REUSE_OBJECT, false)) {
+            return o;
+        }
+        if (o instanceof Writable) {
+            ((Writable) o).write(cloneBufferDataOutput);
+            Writable newO = (Writable) o.getClass().newInstance();
+            newO.readFields(cloneBuffer.getDataInputStream());
+            o = (T) newO;
+            cloneBuffer.reset();
+            return o;
+        }
+        return serializationService.toObject(serializationService.toData(o));
+    }
+
+    private static InputFormat getInputFormat(Configuration configuration) {
+        Class<?> inputFormatClass = configuration.getClass(MRJobConfig.INPUT_FORMAT_CLASS_ATTR, TextInputFormat.class);
+        try {
+            return (InputFormat) ReflectionUtils.newInstance(inputFormatClass, configuration);
```

Comment:
why not just create the instance ourselves, rather than creating and then deleting from cache? It doesn't seem to do any magic.

Reply:
it does do some magic actually, it tries to set `Configuration` object to the created object if it is configurable and it tries to call `configure` method, again using reflection.

Reply:
but we can also call ReflectionUtils.setConf() explicitly after creation.